app/domain/authentication/authn_jwt/validate_and_decode_token.rb

```diff
@@ -0,0 +1,144 @@
+module Authentication
+  module AuthnJwt
+
+    # ValidateAndDecodeToken command class is responsible to validate the JWT token 2 times:
+    # 1st we are validating only the signature.
+    # 2nd we are validating the claims, by checking the token content to decide which claims are mandatory
+    # for the 2nd validation
+    ValidateAndDecodeToken ||= CommandClass.new(
+      dependencies: {
+        fetch_signing_key: ::Util::ConcurrencyLimitedCache.new(
+          ::Util::RateLimitedCache.new(
+            ::Authentication::AuthnJwt::FetchCachedSigningKey.new(
+              singing_key_interface: ::Authentication::AuthnJwt::FetchSigningKeyInterface
+            ),
+            refreshes_per_interval: CACHE_REFRESHES_PER_INTERVAL,
+            rate_limit_interval: CACHE_RATE_LIMIT_INTERVAL,
+            logger: Rails.logger
+          ),
+            max_concurrent_requests: CACHE_MAX_CONCURRENT_REQUESTS,
+            logger: Rails.logger
+        ),
+        verify_and_decode_token: ::Authentication::Jwt::VerifyAndDecodeToken.new,
+        fetch_jwt_claims_to_validate: ::Authentication::AuthnJwt::FetchJwtClaimsToValidate.new,
+        get_verification_option_by_jwt_claim: ::Authentication::AuthnJwt::GetVerificationOptionByJwtClaim.new,
+        logger: Rails.logger
+      },
+      inputs: %i[authentication_parameters]
+    ) do
+      extend(Forwardable)
+      def_delegators(:@authentication_parameters, :jwt_token)
+
+      def call
+        @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatingToken.new)
+        validate_token_exists
+        fetch_signing_key
+        validate_signature
+        fetch_jwt_claims_to_validate
+        validate_claims
+        validate_claims
+        decoded_token_after_claims_validation
+        @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatedToken.new)
+
+        @decoded_token_after_claims_validation
+      end
+
+      private
+
+      def fetch_signing_key(force_read: false)
+        @jwks = @fetch_signing_key.call(refresh: force_read)
+        @logger.debug(LogMessages::Authentication::AuthnJwt::SigningKeysFetchedFromCache.new)
+      end
+
+      def validate_token_exists
+        raise Errors::Authentication::AuthnJwt::MissingToken.new if jwt_token.blank?
+      end
+
+      def validate_signature
+        @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatingTokenSignature.new)
+        ensure_keys_are_fresh
+        decoded_token_after_signature_only_validation
```

Comment:
why call `ensure_keys_are_fresh` while its basically just call `decoded_token_after_signature_only_validation` two times?

Reply:
we have this discussion in previous authenticators and we have decided to create this function which makes the flow more readable